amxmodx/CTask.h

```diff
@@ -10,99 +10,121 @@
+		std::shared_ptr<CForward> m_Forward;
 		cell m_iId;
-		int m_iFunc;
-		int m_iRepeat;
-		
+
+		bool m_bKillMe;
 		bool m_bInExecute;
 		bool m_bLoop;
 		bool m_bAfterStart;
 		bool m_bBeforeEnd;
-		float m_fBase;		// for normal tasks, stores the interval, for the others, stores the amount of time before start / after end
+
+		int m_iRepeat;
 		int m_iParamLen;
-		
-		cell *m_pParams;
-		bool m_bFree;
 
-		// execution
 		float m_fNextExecTime;
+		float m_fBase; // for normal tasks, stores the interval, for the others, stores the amount of time before start / after end
+
 	public:
-		void set(CPluginMngr::CPlugin *pPlugin, int iFunc, int iFlags, cell iId, float fBase, int iParamsLen, const cell *pParams, int iRepeat, float fCurrentTime);
-		void clear();
-		bool isFree() const;
+
+		CTask(CPluginMngr::CPlugin *pPlugin, int iFlags, cell iId, float fBase, int iParamsLen, const cell *pParams, int iRepeat, float fCurrentTime, char * sFunc);
 
 		inline CPluginMngr::CPlugin *getPlugin() const { return m_pPlugin; }
 		inline AMX *getAMX() const { return m_pPlugin->getAMX(); }
 		inline int getTaskId() const { return m_iId; }
+		inline bool shouldKillMe() const { return m_bKillMe; }
+		inline bool inExecute() const { return m_bInExecute; }
+		inline bool isFailed() const { return m_Forward->m_iFunc == -1; }
 
-		void executeIfRequired(float fCurrentTime, float fTimeLimit, float fTimeLeft);	// also removes the task if needed
-
+		void killMe();
 		void changeBase(float fNewBase);
 		void resetNextExecTime(float fCurrentTime);
-		inline bool inExecute() const { return m_bInExecute; }
 
-		bool shouldRepeat();
+		bool execute();
+
+		int executionRequiredStatus(float fCurrentTime, float fTimeLimit, float fTimeLeft);
+
+		friend bool operator < (const CTask &left, const CTask &right)
+		{
+			if (left.m_bAfterStart || left.m_bAfterStart)
+				return true;
 
-		CTask();
-		~CTask();
+			return left.m_fNextExecTime < right.m_fNextExecTime;
+		}
 	};
 
 	class CTaskDescriptor
 	{
 	public:
 		cell m_iId;
 		AMX *m_pAmx;
-		bool m_bFree;
 
-		CTaskDescriptor(int iId, AMX *pAmx, bool bFree = false)
+		CTaskDescriptor(int iId, AMX *pAmx)
 		{
 			m_iId = iId;
 			m_pAmx = pAmx;
-			m_bFree = bFree;
 		}
 
 		friend bool operator == (const CTask &left, const CTaskDescriptor &right)
 		{
-			if (right.m_bFree)
-				return (left.isFree() && !left.inExecute());
-			
-			return (!left.isFree()) && 
-					(right.m_pAmx ? left.getAMX() == right.m_pAmx : true) && 
-					(left.getTaskId() == right.m_iId);
+			// Dead tasks are invalid and may not be compared
+			if (left.shouldKillMe())
+				return false;
+
+			if (right.m_pAmx == 0 || right.m_pAmx == left.getAMX())
+			{
+				if (right.m_iId == left.getTaskId())
+				{
+					return true;
+				}
+			}
+			return false;
 		}
 	};
 
 	/*** CTaskMngr priv members ***/
-	typedef CList<CTask, CTaskDescriptor> TaskList;
-	typedef TaskList::iterator TaskListIter;
-	
-	TaskList m_Tasks;
+	std::vector<CTask> m_Tasks;
```

Comment:
I have no idea about the compatibility part, but comparing to ke::Vector, from some quick benchmarking on MSVC, ke::Vector is consistently slower (around 50% slower for appends and 30% for some mixed workload), there doesn't seem to be a sort algorithm (maybe I'm blind though) which would be necessary here (I think std::sort should work, but then again, you'd be using std), there is no efficient erase remove operation (if you want to remove 100 objects, it's going to be incredibly slow because it will shift everything a 100 times), no move semantics which helps quite a bit with performance too, and whatnot, that's why I went with std. 

Anyway, I'd understand if you want to reject this PR because the std dependency, and don't take this comparison personal or as a std vs amtl war, I'm probably wrong in many of the points I mentioned.